keps/sig-scheduling/5194-reserved-for-workloads/README.md

```diff
@@ -0,0 +1,796 @@
+Removing this limit does not mean that the number of pods that can share a ResourceClaim
+will be unlimited. As part of the
+[scale testing effort for DRA](https://github.com/kubernetes/kubernetes/issues/131198),
+we will test the scalability of the number of pods sharing a ResourceClaim so we can
+provide guidance as to what is a safe number.
+
+## Design Details
+
+### Background
+
+The `ReservedFor` field on the `ResourceClaimStatus` is currently used for two purposes:
+
+#### Deallocation
+Devices are allocated to `ResourceClaim`s when the first pod referencing the claim is
+scheduled. Other pods can also share the `ResourceClaim` in which case they share the
+devices. Once no pods are consuming the claim, the devices should be deallocated to they
+can be allocted to other claims. The `ReservedFor` list is used to keep track of pods
+consuming a `ResourceClaim`. Pods are added to the list by the DRA scheduler plugin
+during scheduling and removed from the list by the resourceclaim controller when pods are
+deleted or finish running. An empty list means there are no current consumers of the claim
+and it can be deallocated.
+
+#### Finding pods using a ResourceClaim
+It is used by the DRA scheduler plugin, the kubelet, and the device_taint_eviction
+controller to find pods that are using the ResourceClaim:
+
+1. The kubelet uses this to make sure it only runs pods that where the claims have been allocated
+   to the pod. It can verify this by checking that the Pod is listed in the `ReservedFor` list.
+
+1. The DRA scheduler plugin uses the list to find claims that have zero or only
+   a single pod using it, and is therefore a candidate for deallocation in the `PostFilter` function.
+
+1. The device_taint_eviction controller uses the `ReservedFor` list to find the pods that needs to be evicted
+   when one or more of the devices allocated to a ResourceClaim is tainted (and the ResourceClaim
+   does not have a toleration).
+
+So the solution needs to:
+* Give the resourceclaim controller a way to know when there are no more consumers of a ResourceClaim so
+  it can be deallocated.
+* Give controllers a way to list the pods consuming or referencing a ResourceClaim.
+
+### Proposal
+
+#### API
+
+The exact set of proposed API changes can be seen below (`...` is used in places where new fields
+are added to existing types):
+
+```go
+// ResourceClaimSpec defines what is being requested in a ResourceClaim and how to configure it.
+type ResourceClaimSpec struct {
+  ...
+
+  // ReservedFor specifies the resource that will be consuming the claim. If set, the
+  // reference will be copied into the status.ReservedFor list when the claim is allocated.
+  //
+  // When this field is set it is the responsibility of the entity that created the
+  // ResourceClaim to remove the reference from the status.ReservedFor list when there
+  // are no longer any pods consuming the claim.
+  //
```

Comment:
Most user-created ResourceClaims should not set this field. It is more typically used by ResourceClaims created and managed by controllers.